module/os/linux/zfs/zfs_ioctl_os.c

```diff
@@ -136,22 +136,18 @@ zfsdev_state_init(struct file *filp)
 	return (0);
 }
 
-static int
-zfsdev_state_destroy(struct file *filp)
+static void
+zfsdev_state_destroy(zfsdev_state_t *zs)
 {
-	zfsdev_state_t *zs;
-
-	ASSERT(MUTEX_HELD(&zfsdev_state_lock));
-	ASSERT(filp->private_data != NULL);
+	ASSERT(zs != NULL);
+	ASSERT3S(zs->zs_minor, >, 0);
 
-	zs = filp->private_data;
-	zs->zs_minor = -1;
 	zfs_onexit_destroy(zs->zs_onexit);
 	zfs_zevent_destroy(zs->zs_zevent);
 	zs->zs_onexit = NULL;
 	zs->zs_zevent = NULL;
-
-	return (0);
+	smp_wmb();
```

Comment:
Wouldn't it be better to migrate to more cross-platform membar_producer() ?

Reply:
That sounds good to me, it will avoid having to define smp_wmb for FreeBSD in the next PR.